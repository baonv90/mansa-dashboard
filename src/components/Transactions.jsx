import React from 'react';
import styled from 'styled-components';


const Container = styled.div`
  display: flex;
  flex-direction: column;
  overflow: auto;
`;

export const Item = styled.div`
  display: flex;
  font-size: .825rem;
  line-height: 1.25rem;
  align-items: center;
  padding: .5rem;
  margin-right: 1rem;
  border-bottom: 1px solid var(--grey-color);
`;

const Header = styled(Item)`
  font-weight: 600;
  border-bottom: 1px solid var(--border-color);
`;

const Info = styled.div`
  width: 25%;
`;

const Amount = styled.div`
  width: 25%;
  display: flex;
  color: ${({ moneyIn, header }) => header 
    ? 'var(--title-color)' 
    : moneyIn ? 'var(--in-color)' : 'var(--out-color)'};
  justify-content: flex-end;
  font-weight: 600;
`;

// Simple text format
const formatText = (str) => {
  return str && str.replace('_', ' ').toLowerCase();
};

// Transaction history
const Transactions = ({ data }) => {
  return (
    <>
      <Header>
        <Info>Date</Info>
        <Info>Type</Info>
        <Info>Category</Info>
        <Amount header={true}>Amount</Amount>
      </Header>
      <Container>
        {(data || []).map((dat, index) => {
          const { timestamp, amount, currency, transaction_category, transaction_type } = dat;
          const displayDate = timestamp.slice(0, timestamp.indexOf(' '));

          return (
            <Item key={`transaction-${index}`}>
              <Info>{displayDate}</Info>
              <Info>{formatText(transaction_type)}</Info>
              <Info>{formatText(transaction_category)}</Info>
              <Amount moneyIn={amount >= 0}>{amount}{currency === 'GBP' ? 'Â£' : currency}</Amount>
            </Item>
          );
        })}
      </Container>
    </>
  );
};

export default Transactions;