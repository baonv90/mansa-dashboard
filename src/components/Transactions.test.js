import renderer from 'react-test-renderer';
import Transactions, { Item } from './Transactions';

describe('Transactions', () => {
  const props = {
    data: [
      {
        timestamp: '2021/08/07 11:40:33',
        amount: 11, 
        currency: 'GBP', 
        transaction_category: 'DEBIT',
        transaction_type: 'ATM'
      }, {
        timestamp: '2021/08/07 11:40:33',
        amount: 11, 
        currency: 'GBP', 
        transaction_category: 'DEBIT',
        transaction_type: 'ATM'
      }
    ]
  };

  it('should match snapshot', () =>  {
    const component = renderer.create(
      <Transactions {...props} />,
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot()
  });

  it('should render correctly number of transactions', () =>  {
    const component = renderer.create(
      <Transactions {...props} />,
    );
    
    const els = component.root.findAllByType(Item);
    expect(els).toHaveLength(props.data.length);
  });
});