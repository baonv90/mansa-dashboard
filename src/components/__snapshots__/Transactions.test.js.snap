// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components > Transactions should match snapshot 1`] = `
Array [
  <div
    className="sc-gtsrHT sc-dlnjwi iJJHtu snPHy"
  >
    <div
      className="sc-hKFxyN bArvyH"
    >
      Date
    </div>
    <div
      className="sc-hKFxyN bArvyH"
    >
      Type
    </div>
    <div
      className="sc-hKFxyN bArvyH"
    >
      Category
    </div>
    <div
      className="sc-eCApnc kDPHZL"
    >
      Amount
    </div>
  </div>,
  <div
    className="sc-bdnxRM krqvSh"
  >
    <div
      className="sc-gtsrHT iJJHtu"
    >
      <div
        className="sc-hKFxyN bArvyH"
      >
        2021/08/07
      </div>
      <div
        className="sc-hKFxyN bArvyH"
      >
        atm
      </div>
      <div
        className="sc-hKFxyN bArvyH"
      >
        debit
      </div>
      <div
        className="sc-eCApnc icuxwU"
      >
        11
        £
      </div>
    </div>
    <div
      className="sc-gtsrHT iJJHtu"
    >
      <div
        className="sc-hKFxyN bArvyH"
      >
        2021/08/07
      </div>
      <div
        className="sc-hKFxyN bArvyH"
      >
        atm
      </div>
      <div
        className="sc-hKFxyN bArvyH"
      >
        debit
      </div>
      <div
        className="sc-eCApnc icuxwU"
      >
        11
        £
      </div>
    </div>
  </div>,
]
`;
